src/sage/rings/cfinite_sequence.py

```diff
@@ -650,28 +649,27 @@ def __getitem__(self, key):
             m = max(key.start, key.stop)
             return [self[ii] for ii in range(*key.indices(m + 1))]
         elif isinstance(key, Integral):
-            from sage.matrix.constructor import Matrix
-            d = self._deg
-            if (self._off <= key and key < self._off + len(self._a)):
-                return self._a[key - self._off]
-            elif d == 0:
+            den = self.denominator()
+            num = self.numerator()
+            if self._off >= 0:
+                num = num.shift(-self._off)
+            else:
+                den = den.shift(self._off)
+            n = key - self._off
+            if n < 0:
                 return 0
-            (quo, rem) = self.numerator().quo_rem(self.denominator())
-            wp = quo[key - self._off]
-            if key < self._off:
-                return wp
-            A = Matrix(QQ, 1, d, self._c)
-            B = Matrix.identity(QQ, d - 1)
-            C = Matrix(QQ, d - 1, 1, 0)
-            if quo == 0:
-                off = self._off
-                V = Matrix(QQ, d, 1, self._a[:d][::-1])
+            (quo, num) = num.quo_rem(den)
+            P = self.parent().polynomial_ring()
+            x = self.parent().gen()
+            if quo.degree() < n:
+                wp = 0
             else:
-                off = 0
-                V = Matrix(QQ, d, 1, self._aa[:d][::-1])
-            M = Matrix.block([[A], [B, C]], subdivide=False)
-
-            return wp + list(M ** (key - off) * V)[d - 1][0]
+                wp = quo[n]
+            while n:
+                num = P((num * den(-x)).list()[n % 2::2])
```

Comment:
`den(-x)` actually takes quite a bit of time to compute in practice with the current implementation, so you may want to do it only once.

Reply:
Thank you. I improved it. In my enviorment, the improvment was very small. I also tried

```python
nden = P([v if i % 2 == 0 else -v for i, v in enumerate(den)])
```

But, it was slower.